### ISSUE TITLE
Deadlock when nesting transactions for different documents?

### ISSUE THREAD

**Issue by SebastianStehle** [y-crdt/y-crdt#329]
Hi,

I am playing around with a dotnet wrapper and my fork. The goal is to implement some business logic in the backend through callbacks of the server. For example whenever someone creates a chat comment we duplicate the entry to another document for notifications:

https://github.com/SebastianStehle/ydotnet/blob/main/Demo/Callback.cs

I have noticed a deadlock when I create a transaction to another document within another document transaction. Is this a known limitation or an issue on my side?

I have solved it eventually with using an extra task (= potentially other thread).

**Comment by Horusiath** [comment]
Yes. In general, all observer callbacks are called when the read-write transaction commits / gets disposed. This means that at the moment they are called, the transaction is still alive and holds "a lock" (figuratively, it's not a mutex) to its document.

Therefore, **observer callback cannot create new transactions to the same document, as the one that observer uses**. This would mess up with the state of the document during commit phase.

In Yrs the transaction that is executing is passed as a callback argument but it can be used for read-only operations, so you could use it (no need to create a new one). There's a possible extension on this idea that would allow it to also support write operations, but it's not implemented yet and we're not sure if this behaviour should be allowed.

The Yrs docs can tell you more about [transaction lifecycle](https://docs.rs/yrs/latest/yrs/#transaction-event-lifecycle).
